### PR TITLE
Add models to persist embed config

### PR DESCRIPTION
**Pull request by TaiWilkin**
## Overview

The embed config is complex and will grow over time as we add features. Rather than serialize to the query string, store the config in the database and link it to the contributor. 
- Adds configuration models
- Adds creation, update and fetching of embedded map configurations via API
     - Note that the nested embedded map fields are deleted and recreated on update of the parent `EmbedConfig`
- Adds embedded map configuration to the data fetched for user profiles

Connects #1299 

## Demo

Screen Shot 2021-04-06 at 11 58 20 AM

## Testing Instructions

* Run `./scripts/manage migrate`
* Run `./scripts/server` and login as a user with an associated contributor
* Navigate to [:8081/api/embed-configs/](http://localhost:8081/api/embed-configs/)
* `POST` a new embed config object
     - [x] The new EmbedConfig should be created successfully
```
     {
	"width": "100",
	"height": "100",
	"color": "pink",
	"font": "serif",
	"show_other_contributor_information": false,
	"embed_fields": [{
		"column_name": "xyz",
		"display_name": "123",
		"visible": true,
		"order": 1
	}]
   }
```
* Navigate to [your profile](http://localhost:6543/profile/8]
     - [ ] Confirm in the network tab that the EmbedConfig is returned with the contributor details
* Navigate to [your new EmbedConfig](http://localhost:8081/api/embed-configs/1/) and update some fields, including the embed_fields
     - [x] The configuration should update successfully
* Attempt to create an embed_field with a duplicate order
     - [x] You should receive an error message
* Attempt to create a new EmbedConfig with the same user
     - [ ] You should receive an error message

## Checklist

- [x] `fixup!` commits have been squashed
- [x] CI passes after rebase
- [x] CHANGELOG.md updated with summary of features or fixes, following [Keep a Changelog](https://keepachangelog.com/en/1.0.0/) guidelines

### CODE REVIEW

**Review by jwalgran — COMMENTED** [review]
The models look good but I found what look like a few required fixes that we need to make to the API.

**Review by jwalgran — COMMENTED** [review]
I tested the permission updates and they look good. There were two lines in the test instructions that failed for me

- Preventing the creation of multiple configs for a single contributor
- Returning the embed config from the profile endpoint

Consider changes similar to these:

```diff
diff --git a/src/django/api/serializers.py b/src/django/api/serializers.py
index fd9bab1..a9ee624 100644
--- a/src/django/api/serializers.py
+++ b/src/django/api/serializers.py
@@ -171,11 +171,13 @@ class UserProfileSerializer(ModelSerializer):
     other_contributor_type = SerializerMethodField()
     facility_lists = SerializerMethodField()
     is_verified = SerializerMethodField()
+    embed_config = SerializerMethodField()
 
     class Meta:
         model = User
         fields = ('id', 'name', 'description', 'website', 'contributor_type',
-                  'other_contributor_type', 'facility_lists', 'is_verified')
+                  'other_contributor_type', 'facility_lists', 'is_verified',
+                  'embed_config')
 
     def get_name(self, user):
         try:
@@ -233,6 +235,12 @@ class UserProfileSerializer(ModelSerializer):
         except Contributor.DoesNotExist:
             return False
 
+    def get_embed_config(self, user):
+        try:
+            return EmbedConfigSerializer(user.contributor.embed_config).data
+        except Contributor.DoesNotExist:
+            return None
+
 
 class FacilityListSummarySerializer(ModelSerializer):
     contributor_id = SerializerMethodField()
diff --git a/src/django/api/views.py b/src/django/api/views.py
index 5c60122..e5aac11 100644
--- a/src/django/api/views.py
+++ b/src/django/api/views.py
@@ -3534,6 +3534,8 @@ class EmbedConfigViewSet(mixins.ListModelMixin,
             return Response(status=status.HTTP_401_UNAUTHORIZED)
 
         contributor = self.get_contributor(request)
+        if contributor.embed_config is not None:
+            raise ValidationError('Contributor has an existing embed configuration.')
 
         fields_data = request.data.pop('embed_fields', [])
 

```

File: src/django/api/views.py [hunk 1]
```diff
@@ -3489,3 +3492,72 @@ def list(self, request):
...
+    def get_link(self, path, method, base_url):
+        return None
+
+
+def create_embed_fields(fields_data, embed_config):
+    for field_data in fields_data:
+        try:
+            EmbedField.objects.create(embed_config=embed_config, **field_data)
+        except IntegrityError:
+            raise ValidationError('Fields cannot have the same order.')
```
[the remark this answers is not in the document]
Comment (TaiWilkin) [reply src/django/api/views.py]: Updated in 99a9e2c

File: src/django/api/views.py [hunk 2]
```diff
@@ -3489,3 +3492,72 @@ def list(self, request):
...
+            contributor_id = request.user.contributor.id
+            contributor = Contributor.objects.get(id=contributor_id)
+        except Contributor.DoesNotExist:
+            raise ValidationError('Contributor not found for requesting user.')
+
+        if contributor.embed_config is not None:
+            raise ValidationError(
+                'Contributor has an existing embedded map configuration.')
+
+        fields_data = request.data.pop('embed_fields')
```
[the remark this answers is not in the document]
Comment (TaiWilkin) [reply src/django/api/views.py]: Updated in 99a9e2c

File: src/django/oar/urls.py [hunk 3]
```diff
@@ -79,7 +80,8 @@
     path(r'tile/<layer>/<cachekey>/<int:z>/<int:x>/<int:y>.<ext>',
          views.get_tile, name='tile'),
     url(r'^api/current_tile_cache_key', views.current_tile_cache_key),
-    url(r'api-blocks', views.ApiBlockViewSet, 'api-block')
+    url(r'api-blocks', views.ApiBlockViewSet, 'api-block'),
+    url(r'embed-configs', views.EmbedConfigViewSet, 'embed-config')
```
[the remark this answers is not in the document]
Comment (TaiWilkin) [reply src/django/oar/urls.py]: Updated in 99a9e2c

File: src/django/api/views.py [hunk 4]
```diff
@@ -3489,3 +3492,72 @@ def list(self, request):
...
+    """
+    View EmbedConfig.
+    """
+    queryset = EmbedConfig.objects.all()
+    serializer_class = EmbedConfigSerializer
+
+    @transaction.atomic
+    def create(self, request):
+        try:
+            contributor_id = request.user.contributor.id
```
[the remark this answers is not in the document]
Comment (TaiWilkin) [reply src/django/api/views.py:3524]: Updated in 99a9e2c

File: src/django/api/views.py [hunk 5]
```diff
@@ -3489,3 +3492,72 @@ def list(self, request):
...
+            response_data = self.get_serializer(embed_config).data
+
+            return Response(response_data)
+
+    @transaction.atomic
+    def update(self, request, pk=None):
+        fields_data = request.data.pop('embed_fields')
+
+        embed_config = EmbedConfig.objects.get(id=pk)
+
```
[the remark this answers is not in the document]
Comment (TaiWilkin) [reply src/django/api/views.py:3565]: Updated in 99a9e2c

**Comment by TaiWilkin** [comment]
Thanks for the feedback! I've updated the profile fetching and the contributor check in d4b0b0c

**Review by jwalgran — APPROVED** [review]
Looks good. Thanks for working through the changes with me.

**Comment by TaiWilkin** [comment]
Thank you for reviewing!

**Pull request merged**